### ISSUE TITLE
OpenAPI spec generated by Documentation plugin in v4 no longer creates component schemas

### ISSUE THREAD

**Issue by djvaldez** [strapi/strapi#12297]
## Bug report

As of Strapi v4.0.5, the Documentation plugin does not create component schemas in the OpenAPI spec. It's not clear if this was purposely omitted, but v3.6.8 produces top-level schema definitions in the spec instead of in-line response definitions. This limits the usefulness of code generation, e.g. OpenAPI Typescript Codegen won't generate models from the schema generated by v4.0.5 but will generate models for the spec created by v3.6.8.

### Steps to reproduce the behavior

1. Install the Documentation plugin and run the Strapi instance.
2. Go to the default documentation page (`/documentation/v1.0.0`).
3. Observe that there is no "Schemas" section on the page
4. Alternatively, check the generated spec in the project directory: `src/extensions/documentation/documentation/1.0.0/full_documentation.json`. The generated spec for v3 contains the root-level object `components`; the v4 version does not.

### Expected behavior

Expected behavior compared to v3 is to see the schema definitions listed on the page, and schema definitions at the root level of  `full_documentation.json` instead of in-line defined responses.

### Screenshots

v3 Documentation page:
![v3 documentation](https://user-images.githubusercontent.com/1976701/151027659-d1a2a497-d063-4859-9407-4cfce75e7132.png)

v4 Documentation page:
![v4 documentation](https://user-images.githubusercontent.com/1976701/151027741-25e3ea3f-ff13-424b-9bc4-061ab1c04ab2.png)

v3 response (observe the reference to the "Product" definition)
![v3 response](https://user-images.githubusercontent.com/1976701/151027836-317c0fda-c6d0-4ee9-9dc9-9b62c16556ff.png)

v4 response (definition is in-line):
![v4 response](https://user-images.githubusercontent.com/1976701/151027867-23cedcae-e9d7-42a5-8fa5-f9940ebcb973.png)

### Code snippets
Some relevant changes to get basic schemas generated for response entities for v4 are included below. Other changes are needed to support these. In this example, list responses generate a schema object with a name in the form of`<ContentTypeName>ListResponse` that contains an array inside the object; however, in v3 list responses were simple arrays of the content schema.

In `plugins/documentation/server/utils/builders/build-api-endpoint-path.js`:
```diff
+  const schemas = {};
   const paths = routeInfo.routes.reduce((acc, route) => {
     // TODO: Find a more reliable way to determine list of entities vs a single entity
     const isListOfEntities = route.handler.split('.').pop() === 'find';
     const methodVerb = route.method.toLowerCase();
     const hasPathParams = route.path.includes('/:');
     const pathWithPrefix = routeInfo.prefix
       ? getPathWithPrefix(routeInfo.prefix, route.path)
       : route.path;
     const routePath = hasPathParams ? parsePathWithVariables(pathWithPrefix) : pathWithPrefix;

-    const { responses } = buildApiResponses(attributes, route, isListOfEntities);
+    const { responses } = buildApiResponses(tag, route, isListOfEntities);
+    for (const response of _.values(responses)) {
+      for (const contentTypeResponse of _.values(response.content)) {
+        if (_.has(contentTypeResponse.schema, '$ref')) {
+          const schemaName = _.last(_.split(contentTypeResponse.schema['$ref'], '/'));
+          schemas[schemaName] = {
+            type: 'object',
+            properties: {
+              data: getSchemaData(isListOfEntities, cleanSchemaAttributes(attributes)),
+              meta: getMeta(isListOfEntities),
+            },
+          };
+        }
+      }
+    }

     const swaggerConfig = {
       responses,
@@ -120,7 +163,7 @@ const getPaths = ({ routeInfo, attributes, tag }) => {
     return acc;
   }, {});

-  return { paths };
+  return { paths, schemas };
 };
```
In `packages/plugins/documentation/server/utils/builders/build-api-responses.js`:
```diff
-module.exports = (attributes, route, isListOfEntities = false) => {
+module.exports = (tag, route, isListOfEntities = false) => {
   let schema;
   if (route.method === 'DELETE') {
     schema = {
@@ -46,11 +19,11 @@ module.exports = (attributes, route, isListOfEntities = false) => {
       format: 'int64',
     };
   } else {
+    const responseName = `${_.words(_.startCase(tag)).join('')}${
+      isListOfEntities ? 'List' : ''
+    }Response`;
     schema = {
-      properties: {
-        data: getSchemaData(isListOfEntities, cleanSchemaAttributes(attributes)),
-        meta: getMeta(isListOfEntities),
-      },
+      $ref: `#/components/schemas/${responseName}`,
     };
   }
```

### System

- Node.js version: v16.13.1
- NPM version: 8.1.2
- Strapi version: v4.0.5
- Database:
- Operating system:

**Comment by emilieringwald** [comment]
Hello @djvaldez! Our team will have a look at this in the coming weeks, we'll keep you posted.

**Comment by markkaylor** [comment]
@djvaldez Thanks for submitting this issue. To migrate this plugin to v4 we basically started from scratch. Due to time constraints some things had to be omitted and this was one of them. Sorry about that, we will be dedicating time to this plugin soon and this feature will be added back.

**Comment by DimitriGilbert** [comment]
Hi,
Starting from @djvaldez code I added a few things like :
* components for response data
* response schemas
* relation ref

You can find the code here : https://github.com/DimitriGilbert/strapi/tree/oa_doc_generation

it is certainly not pretty but you get a good idea and won't have to browse a huge amount of files (mostly packages/plugins/documentation/server/services/documentation.js).
I hope this helps :+1:

**Comment by DennisHartrampf** [comment]
The solution to this issue would be of great benefit for my team. Right now we are heavily creating new content types in Strapi during our sprints. And at the moment we need to manually implement the DTOs and services in our frontend Angular app(s). It would be fantastic to use the OpenAPI generator to auto-generate the DTOs and services for use in our frontend code out of the OpenAPI spec - like we do for all our other APIs.

@markkaylor or @emilieringwald do you have any estimation as to when this issue can be expected to be fixed?

Thanks in advance!

**Comment by markkaylor** [comment]
Hi @DennisHartrampf we will try to get some of these Documentation issues and PR reviews in our next sprint.

**Comment by DennisHartrampf** [comment]
Hi @markkaylor, thanks for the estimate.

**Comment by JiRadu** [comment]
I'm in the same situation a @Hartrampf , we really would appriciate this functionality in order to use it in FE.
I've tried the code from @DimitriGilbert but it's pretty difficult to make modifications to it and it's a bit messy.
We decided to wait for the official release of the new version of this plugin.
@markkaylor how long are your sprints - 2weeks or 1 month or something else?

Thank you for taking this into consideration!
